File: itsm-frontend/src/types/ticket-type.ts
/**
 * 工单类型管理类型定义
 */

import { ApprovalNodeConfig } from './workflow';

/**
 * 工单类型状态
 */
export enum TicketTypeStatus {
  ACTIVE = 'active',
  INACTIVE = 'inactive',
  DRAFT = 'draft',
}

/**
 * 自定义字段类型
 */
export enum CustomFieldType {
  TEXT = 'text',
  TEXTAREA = 'textarea',
  NUMBER = 'number',
  DATE = 'date',
  DATETIME = 'datetime',
  SELECT = 'select',
  MULTI_SELECT = 'multi_select',
  CHECKBOX = 'checkbox',
  RADIO = 'radio',
  FILE = 'file',
  USER_PICKER = 'user_picker',
  DEPARTMENT_PICKER = 'department_picker',
}

/**
 * 自定义字段定义
 */
export interface CustomFieldDefinition {
  id: string;
  name: string;
  label: string;
  type: CustomFieldType;
  required: boolean;
  description?: string;
  placeholder?: string;
  defaultValue?: any;
  options?: Array<{
    label: string;
    value: string | number;
  }>;
  validation?: {
    min?: number;
    max?: number;
    pattern?: string;
    message?: string;
  };
  conditionalDisplay?: {
    field: string;
    operator: 'equals' | 'not_equals' | 'contains';
    value: any;
  };
  order: number;
}

/**
 * 工单类型定义
 */
export interface TicketTypeDefinition {
  id: number;
  code: string;
  name: string;
  description?: string;
  icon?: string;
  color?: string;
  status: TicketTypeStatus;
  
  // 表单配置
  customFields: CustomFieldDefinition[];
  
  // 审批流程
  approvalEnabled: boolean;
  approvalWorkflowId?: string;
  approvalChain?: ApprovalChainDefinition[];
  
  // SLA配置
  slaEnabled: boolean;
  defaultSlaId?: number;
  
  // 分配规则
  autoAssignEnabled: boolean;
  assignmentRules?: AssignmentRule[];
  
  // 通知配置
  notificationConfig?: NotificationConfig;
  
  // 权限配置
  permissionConfig?: PermissionConfig;
  
  // 元数据
  createdBy: number;
  createdByName: string;
  updatedBy?: number;
  updatedByName?: string;
  createdAt: string;
  updatedAt: string;
  tenantId: number;
  departmentId?: number; // 部门ID
  
  // 统计
  usageCount?: number;
}

/**
 * 审批链定义
 */
export interface ApprovalChainDefinition {
  id: string;
  level: number;
  name: string;
  approvers: Array<{
    type: 'user' | 'role' | 'department' | 'dynamic';
    value: number | string;
    name: string;
  }>;
  approvalType: 'any' | 'all' | 'majority';
  minimumApprovals?: number;
  allowReject: boolean;
  allowDelegate: boolean;
  rejectAction: 'end' | 'return' | 'custom';
  returnToLevel?: number;
  conditions?: Array<{
    field: string;
    operator: 'equals' | 'not_equals' | 'greater_than' | 'less_than';
    value: any;
  }>;
  timeout?: number; // 超时时间（小时）
  timeoutAction?: 'auto_approve' | 'auto_reject' | 'escalate';
}

/**
 * 分配规则
 */
export interface AssignmentRule {
  id: string;
  name: string;
  priority: number;
  conditions: Array<{
    field: string;
    operator: 'equals' | 'not_equals' | 'contains' | 'greater_than' | 'less_than';
    value: any;
  }>;
  assignTo: {
    type: 'user' | 'role' | 'department' | 'round_robin' | 'load_balance';
    value: number | string;
  };
  enabled: boolean;
}

/**
 * 通知配置
 */
export interface NotificationConfig {
  onCreate: {
    enabled: boolean;
    recipients: Array<'requester' | 'assignee' | 'approvers' | 'watchers'>;
    template?: string;
  };
  onUpdate: {
    enabled: boolean;
    recipients: Array<'requester' | 'assignee' | 'approvers' | 'watchers'>;
  };
  onApproval: {
    enabled: boolean;
    recipients: Array<'requester' | 'assignee' | 'previous_approvers' | 'next_approvers'>;
  };
  onReject: {
    enabled: boolean;
    recipients: Array<'requester' | 'assignee' | 'approvers'>;
  };
  onComplete: {
    enabled: boolean;
    recipients: Array<'requester' | 'assignee' | 'watchers'>;
  };
}

/**
 * 权限配置
 */
export interface PermissionConfig {
  canCreate: string[]; // 角色列表
  canView: string[];
  canEdit: string[];
  canDelete: string[];
  canApprove: string[];
}

/**
 * 创建工单类型请求
 */
export interface CreateTicketTypeRequest {
  code: string;
  name: string;
  description?: string;
  icon?: string;
  color?: string;
  customFields: CustomFieldDefinition[];
  approvalEnabled: boolean;
  approvalChain?: ApprovalChainDefinition[];
  slaEnabled: boolean;
  defaultSlaId?: number;
  autoAssignEnabled: boolean;
  assignmentRules?: AssignmentRule[];
  notificationConfig?: NotificationConfig;
  permissionConfig?: PermissionConfig;
  departmentId?: number;
}

/**
 * 工单类型列表响应
 */
export interface UpdateTicketTypeRequest {
  name?: string;
  description?: string;
  icon?: string;
  color?: string;
  status?: TicketTypeStatus;
  customFields?: CustomFieldDefinition[];
  approvalEnabled?: boolean;
  approvalChain?: ApprovalChainDefinition[];
  slaEnabled?: boolean;
  defaultSlaId?: number;
  autoAssignEnabled?: boolean;
  assignmentRules?: AssignmentRule[];
  notificationConfig?: NotificationConfig;
  permissionConfig?: PermissionConfig;
  departmentId?: number;
}

/**
 * 工单类型列表响应
 */
export interface TicketTypeListResponse {
  types: TicketTypeDefinition[];
  total: number;
  page: number;
  pageSize: number;
}

File: itsm-backend/ent/schema/ticket_type.go
package schema

import (
	"entgo.io/ent"
	"entgo.io/ent/schema/edge"
	"entgo.io/ent/schema/field"
	"entgo.io/ent/schema/index"
)

// TicketType holds the schema definition for the TicketType entity.
type TicketType struct {
	ent.Schema
}

// Fields of the TicketType.
func (TicketType) Fields() []ent.Field {
	return []ent.Field{
		field.String("code").NotEmpty().MaxLen(50),
		field.String("name").NotEmpty().MaxLen(100),
		field.Text("description"),
		field.String("icon").MaxLen(50),
		field.String("color").MaxLen(20),
		field.String("status").Default("active"),
		field.Int("category_id").Optional(),
		field.String("default_priority").Default("medium").MaxLen(20),
		field.Int("sort_order").Default(0),
		field.String("workflow_definition_key").Optional().MaxLen(128),
		field.Int("assignment_rule_id").Optional(),
		field.Time("archived_at").Optional().Nillable(),
		field.Int64("archived_by").Optional(),
		field.JSON("custom_fields", map[string]interface{}{}),
		field.Bool("approval_enabled").Default(false),
		field.Int64("approval_workflow_id").Optional(),
		field.JSON("approval_chain", []interface{}{}),
		field.Bool("sla_enabled").Default(false),
		field.Int64("default_sla_id").Optional(),
		field.Bool("auto_assign_enabled").Default(false),
		field.JSON("assignment_rules", []interface{}{}),
		field.JSON("notification_config", map[string]interface{}{}),
		field.JSON("permission_config", map[string]interface{}{}),
		field.Int64("tenant_id"),
		field.Int64("created_by"),
		field.Time("created_at"),
		field.Time("updated_at"),
		field.Int64("updated_by").Optional(),
		field.Int("usage_count").Default(0),
	}
}

// Edges of the TicketType.
func (TicketType) Edges() []ent.Edge {
	return []ent.Edge{edge.To("tickets", Ticket.Type)}
}

// Indexes of the TicketType.
func (TicketType) Indexes() []ent.Index {
	return []ent.Index{
		index.Fields("code", "tenant_id").Unique(),
		index.Fields("tenant_id"),
		index.Fields("status"),
		index.Fields("tenant_id", "sort_order"),
	}
}
